File: release2/library/asful.f
C***********************************************************************
C$SPLIT$ASFUL$*********************************************************
C***********************************************************************
      SUBROUTINE ASFUL(SYSK, ISYSK, JSYSK, ELK, IELK, JELK, STEER,
     *     ISTEER, DOFEL, ITEST)
C-----------------------------------------------------------------------
C PURPOSE
C      ASSEMBLES FULL REAL SYSTEM MATRIX
C
C HISTORY
C
C      COPYRIGHT (C) 1979 : SERC, RUTHERFORD APPLETON LABORATORY
C                           CHILTON, DIDCOT, OXFORDSHIRE OX11 0QX
C
C      RELEASE 1.1  29 OCT 1979 (CG)
C      COMMENTED     6 FEB 1980 (KR)
C
C ARGUMENTS IN
C      SYSK    CONTAINS SYSTEM MATRIX PRIOR TO ADDITION OF
C              CURRENT ELEMNT MATRIX CONTRIBUTION
C      ISYSK   FIRST DIMENSION OF SYSK (.GE. TOTAL NUMBER OF
C              UNCONSTRAINED DEGREES OF FREEDOM)
C      JSYSK   SECOND DIMENSION OF SYSK (.GE. TOTAL NUMBER OF
C              UNCONSTRAINED DEGREES OF FREEDOM)
C      ELK     ELEMENT MATRIX
C      IELK    FIRST DIMENSION OF ELK (.GE. DOFEL)
C      JELK    SECOND DIMENSION OF ELK (.GE. DOFEL)
C      STEER   CONTAINS FREEDOM NUMBERS ASSOCIATED WITH ELEMENT
C              MATRIX CONTRIBUTIONS TO SYSTEM MATRIX
C      ISTEER  DIMENSION OF STEER (.GE. DOFEL)
C      DOFEL   MAXIMUM NUMBER OF DEGREES OF FREEDOM ASSOCIATED
C              WITH ELEMENT TYPE
C      ITEST   ERROR CHECKING OPTION
C
C ARGUMENTS OUT
C      SYSK    SYSTEM MATRIX
C
C ROUTINES CALLED
C      ERRMES
C
C
C     SUBROUTINE ASFUL(SYSK, ISYSK, JSYSK, ELK, IELK, JELK,
C    *     STEER, ISTEER, DOFEL, ITEST)
C***********************************************************************
C
      INTEGER DOFEL, ERRMES, I, IELK, IERROR, ISTEER, ISYSK,
     *     ITEST, J, JELK, JSYSK, JTEST, STEER, STEERI, STEERJ
      DOUBLE PRECISION ELK, SRNAME, SYSK
      DIMENSION ELK(IELK,JELK), STEER(ISTEER),
     *     SYSK(ISYSK,JSYSK)
      DATA SRNAME /8H ASFUL  /
C
C     PARAMETER CHECKING
C
      JTEST = ITEST
      IF (JTEST.EQ.-1) GO TO 1010
      IERROR = 0
      IF (ISTEER.LT.DOFEL) IERROR = 3
      IF (IELK.LT.DOFEL .OR. JELK.LT.DOFEL) IERROR = 2
      IF (DOFEL.LE.0) IERROR = 1
      ITEST = ERRMES(JTEST,IERROR,SRNAME)
      IF (ITEST.NE.0) RETURN
C
C     MAIN LOOPS
C
 1010 DO 1040 I=1,DOFEL
         STEERI = STEER(I)
         IF (STEERI.LE.0) GO TO 1040
         DO 1030 J=1,DOFEL
            STEERJ = STEER(J)
            IF (STEERJ.LE.0) GO TO 1030
C
C     RANGE CHECKING ON STEERI AND STEERJ
C
            IF (JTEST.EQ.-1) GO TO 1020
            IERROR = 0
            IF (ISYSK.LT.STEERI .OR. JSYSK.LT.STEERJ) IERROR = 4
            ITEST = ERRMES(JTEST,IERROR,SRNAME)
            IF (ITEST.NE.0) RETURN
C
 1020       SYSK(STEERI,STEERJ) = SYSK(STEERI,STEERJ) + ELK(I,J)
C
 1030    CONTINUE
 1040 CONTINUE
C
      RETURN
      END

File: felib-c/library-c/asful.cpp
#include "f2c.h"


int asful_(doublereal *sysk, integer *isysk, integer *jsysk, doublereal *elk, 
		   integer *ielk, integer *jelk, integer *steer, integer *isteer, 
		   integer *dofel, integer *itest)
{

	integer errmes_(integer *itest, integer *ierror, char *srname, ftnlen srname_len);

    /* Initialized data */

    static char srname[6+1] = "ASFUL ";

    /* System generated locals */
    integer elk_dim1, elk_offset, sysk_dim1, sysk_offset, i__1, i__2;

    /* Local variables */
    static integer i, j, jtest, steeri, steerj;
    extern integer errmes_();
    static integer ierror;

/* 
-----------------------------------------------------------------------
 PURPOSE

      ASFUL assembles full real system matrix 

 HISTORY 

      Copyright (C) 2000 : CCLRC, Rutherford Appleton Laboratory
                           Chilton, Didcot, Oxfordshire OX11 0QX 

      Release 1.1    29 Oct 1978 (CG)
      Commented       6 Feb 1980 (KR)
      Release 4.0     2 Oct 1999 (CG)
      Release 1.0 (C) 1 May 2000 (CG)

 ARGUMENTS in 
      SYSK    contains system matrix prior to addition of 
              current elemnt matrix contribution 
      ISYSK   first dimension of SYSK (.GE. total number of 
              unconstrained degrees of freedom) 
      JSYSK   second dimension of SYSK (.GE. total number of 
              unconstrained degrees of freedom) 
      ELK     element matrix 
      IELK    first dimension of ELK (.GE. DOFEL) 
      JELK    second dimension of ELK (.GE. DOFEL) 
      STEER   contains freedom numbers associated with element 
              matrix contributions to system matrix 
      ISTEER  dimension of STEER (.GE. DOFEL) 
      DOFEL   maximum number of degrees of freedom associated 
              with element type 
      ITEST   error checking option 

 ARGUMENTS out 
      SYSK    system matrix 

 ROUTINES called 
      ERRMES 

***********************************************************************
*/
/* Parameter adjustments */
    --steer;
    elk_dim1 = *ielk;
    elk_offset = elk_dim1 + 1;
    elk -= elk_offset;
    sysk_dim1 = *isysk;
    sysk_offset = sysk_dim1 + 1;
    sysk -= sysk_offset;

    /* Function Body */

/*     Parameter checking */

    jtest = *itest;
    if (jtest != -1) {
	ierror = 0;
	if (*isteer < *dofel) {
	    ierror = 3;
	}
	if (*ielk < *dofel || *jelk < *dofel) {
	    ierror = 2;
	}
	if (*dofel <= 0) {
	    ierror = 1;
	}
	*itest = errmes_(&jtest, &ierror, srname, 6L);
	if (*itest != 0) {
	    return 0;
	}
    }

/*     Main loops */

    i__1 = *dofel;
    for (i = 1; i <= i__1; ++i) {
	steeri = steer[i];
	if (steeri > 0) {
	    i__2 = *dofel;
	    for (j = 1; j <= i__2; ++j) {
		steerj = steer[j];
		if (steerj > 0) {

/*     Range checking on STEERI and STEERJ */

		    if (jtest != -1) {
			ierror = 0;
			if (*isysk < steeri || *jsysk < steerj) {
			    ierror = 4;
			}
			*itest = errmes_(&jtest, &ierror, srname, 6L);
			if (*itest != 0) {
			    return 0;
			}
		    }
		    sysk[steeri + steerj * sysk_dim1] += elk[i + j * elk_dim1];
		}
	    }
	}
    }
return 0;
} 
/* asful_ */

